Avoid calling SynchronousLoadData CompletionHandler more than once

sysrqb opened the pull request:
#### e336a8ca8c960b36c8a74aece5e5a16d08274960
<pre>
Avoid calling SynchronousLoadData CompletionHandler more than once
<a href="https://bugs.webkit.org/show_bug.cgi?id=270030">https://bugs.webkit.org/show_bug.cgi?id=270030</a>
<a href="https://rdar.apple.com/109257007">rdar://109257007</a>

Reviewed by Chris Dumez.

In some situations, a NetworkResourceLoader may call its SynchronousLoadData&apos;s
delayedReply CompletionHandler more than once. This change returns early if we
hit that case. I&apos;m leaving the debug assertion so we investigate if this starts
happening frequently.

* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::sendReplyToSynchronousRequest):

Canonical link: <a href="https://commits.webkit.org/275465@main">https://commits.webkit.org/275465@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/23aefccf9c05350aa6aec6d11934dc4ee29d6c19

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41765 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20779 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44147 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44342 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37858 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23921 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18110 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34520 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42339 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17710 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35966 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15189 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15401 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45736 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37954 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37312 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41065 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16580 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13614 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39506 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18199 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9393 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18257 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17843 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -125,6 +129,7 @@ static void sendReplyToSynchronousRequest(NetworkResourceLoader::SynchronousLoad
 
     data.delayedReply(data.error, data.response, WTFMove(responseBuffer));
     data.delayedReply = nullptr;
```
This is not necessary, CompletionHandlers already get nulled out when called.

cdumez commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -125,6 +129,7 @@ static void sendReplyToSynchronousRequest(NetworkResourceLoader::SynchronousLoad
 
     data.delayedReply(data.error, data.response, WTFMove(responseBuffer));
     data.delayedReply = nullptr;
+    data.wasCompletionHandlerCalled = true;
```
Not necessary.

cdumez commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -106,6 +106,7 @@ struct NetworkResourceLoader::SynchronousLoadData {
     {
         ASSERT(delayedReply);
     }
+    bool wasCompletionHandlerCalled { false };
```
Not necessary.

cdumez commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -117,6 +118,9 @@ static void sendReplyToSynchronousRequest(NetworkResourceLoader::SynchronousLoad
     ASSERT(data.delayedReply);
     ASSERT(!data.response.isNull() || !data.error.isNull());
 
+    if (data.wasCompletionHandlerCalled)
```
Could just check `if (!data.delayedReply)`.

That said, I am kind of curious how that happens.

sysrqb replied:
Hrm, good point. I wanted to maintain the invariant `ASSERT(data.delayedReply)`, but obviously I didn't actually do that because we already null it.

sysrqb committed:
Avoid calling SynchronousLoadData CompletionHandler more than once
https://bugs.webkit.org/show_bug.cgi?id=270030
rdar://109257007

Reviewed by Chris Dumez.

In some situations, a NetworkResourceLoader may call its SynchronousLoadData's
delayedReply CompletionHandler more than once. This change returns early if we
hit that case. I'm leaving the debug assertion so we investigate if this starts
happening frequently.

* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::sendReplyToSynchronousRequest):

Canonical link: https://commits.webkit.org/275465@main
